# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
# Import your libraries:

%matplotlib inline

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sqlalchemy import create_engine


# Challenge 0 - Load,Query and Create connection of your Dataset

#### A lot of the times you won't have files already saved in an Excel or CSV file for you to prepare your data, implying that a lot of times you'll be extracting data from a Data Warehouse or Data Lake, majority of times through SQL.
#### A couple of times you may want to do some queries on your table in mySQL to get a slight view on what you have in hands,
#### so let's simulate that!

#### First we'll need to create a database and table in mySQL:

##### 1º- Open the austin_weather.sql file in MySQL Workbench and run the script into a desired schema.

#### 2º- As we are in mySQL Workbench, we can do some queries there, to get an overview on some characteristics of our data:
 - a) How many days are recorded in the dataset?
 - b) What is the day with the Highest Temperature in Fahrenheit (column TempHighF)
 - c) What is the average Humidity across all days? (column HumidityAvgPercent)
 - d) Top 10 days, where SeaLevelPressureAvgInches is the highest, knowing DewPointAvgF is higher than 28 ?

In [ ]:
# Write your answers below:

# a) SELECT COUNT(DISTINCT `Date`) AS NumberOfDaysRecorded
FROM austin_weather;
# b)SELECT `Date`, `TempHighF`
FROM austin_weather
ORDER BY `TempHighF` DESC
LIMIT 1;
# c)SELECT AVG(CAST(`HumidityAvgPercent` AS UNSIGNED)) AS AverageHumidity
FROM austin_weather
WHERE `HumidityAvgPercent` REGEXP '^[0-9]+$';
# d)SELECT `Date`, `SeaLevelPressureAvgInches`
FROM austin_weather
WHERE `DewPointAvgF` > 28 AND `SeaLevelPressureAvgInches` REGEXP '^[0-9]+(\.[0-9]+)?$'
ORDER BY CAST(`SeaLevelPressureAvgInches` AS DECIMAL(5,2)) DESC
LIMIT 10;

#### Now that you explored a couple of elements in your table, let's bring your table into this jupyter notebook, by creating a Python-SQL connection like you did on MySQL Project!
#### In case you need a little refresher check this [link](https://www.dataquest.io/blog/sql-insert-tutorial/).
##### 1º - Create a connection using sqlalchemy from python to mysql 
##### 2º- Load the table into a variable called weather_df


In [23]:
# Database connection parameters
db_user = 'root'
db_password = 'pass123'
db_host = 'localhost'  # Or your database host
db_port = '3306'  # Default MySQL port
db_name = 'austin_weather'

# Create connection string using SQLAlchemy
connection_string = f"mysql+pymysql://{db_user}:{db_password}@{db_host}:{db_port}/{db_name}"
engine = create_engine(connection_string)

# Load the table into a DataFrame
query = "SELECT * FROM austin_weather"
weather_df = pd.read_sql(query, con=engine)

# Display the first few rows
print(weather_df.head())

# Close the connection
engine.dispose()


        Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHigh


# Challenge 1 - Describe the Dataset

#### Describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [24]:
# Your code here
# Display column names and data types
print(weather_df.dtypes)

Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                         object
DewPointAvgF                          object
DewPointLowF                          object
HumidityHighPercent                   object
HumidityAvgPercent                    object
HumidityLowPercent                    object
SeaLevelPressureHighInches            object
SeaLevelPressureAvgInches             object
SeaLevelPressureLowInches             object
VisibilityHighMiles                   object
VisibilityAvgMiles                    object
VisibilityLowMiles                    object
WindHighMPH                           object
WindAvgMPH                            object
WindGustMPH                           object
PrecipitationSumInches                object
Events                                object
dtype: object


In [7]:
# Your code here
# Summary statistics for numeric variables
print(weather_df.describe())


                      Date    TempHighF     TempAvgF     TempLowF
count                 1319  1319.000000  1319.000000  1319.000000
mean   2015-10-11 00:00:00    80.862775    70.642911    59.902957
min    2013-12-21 00:00:00    32.000000    29.000000    19.000000
25%    2014-11-15 12:00:00    72.000000    62.000000    49.000000
50%    2015-10-11 00:00:00    83.000000    73.000000    63.000000
75%    2016-09-04 12:00:00    92.000000    83.000000    73.000000
max    2017-07-31 00:00:00   107.000000    93.000000    81.000000
std                    NaN    14.766523    14.045904    14.190648


In [25]:
# Your code here
# View first five rows of the dataset
print(weather_df.head())


        Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHigh

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [26]:
# Your code here
# Checking unique values in the DewPointAvgF column
unique_values = weather_df['DewPointAvgF'].unique()
print(unique_values)

['49' '36' '27' '28' '40' '39' '41' '26' '42' '22' '48' '32' '8' '11' '45'
 '55' '61' '37' '47' '25' '23' '20' '33' '30' '29' '17' '14' '13' '54'
 '59' '15' '24' '34' '35' '57' '50' '53' '60' '46' '56' '51' '31' '38'
 '62' '43' '63' '64' '67' '66' '58' '70' '68' '65' '69' '71' '72' '-' '73'
 '74' '21' '44' '52' '12' '75' '76' '18']


In [ ]:
# Your observation here
Presence of Non-Numeric Values


The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [27]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [28]:
# Convert columns to numeric using apply and pd.to_numeric with error coercion
weather_df[wrong_type_columns] = weather_df[wrong_type_columns].apply(pd.to_numeric, errors='coerce')

# Verify the data types after conversion
print(weather_df.dtypes)



Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                        float64
DewPointAvgF                         float64
DewPointLowF                         float64
HumidityHighPercent                  float64
HumidityAvgPercent                   float64
HumidityLowPercent                   float64
SeaLevelPressureHighInches           float64
SeaLevelPressureAvgInches            float64
SeaLevelPressureLowInches            float64
VisibilityHighMiles                  float64
VisibilityAvgMiles                   float64
VisibilityLowMiles                   float64
WindHighMPH                          float64
WindAvgMPH                           float64
WindGustMPH                          float64
PrecipitationSumInches               float64
Events                                object
dtype: object


#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [29]:
# Your code here
print(weather_df.dtypes)


Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                        float64
DewPointAvgF                         float64
DewPointLowF                         float64
HumidityHighPercent                  float64
HumidityAvgPercent                   float64
HumidityLowPercent                   float64
SeaLevelPressureHighInches           float64
SeaLevelPressureAvgInches            float64
SeaLevelPressureLowInches            float64
VisibilityHighMiles                  float64
VisibilityAvgMiles                   float64
VisibilityLowMiles                   float64
WindHighMPH                          float64
WindAvgMPH                           float64
WindGustMPH                          float64
PrecipitationSumInches               float64
Events                                object
dtype: object


# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [30]:
# Your code here
# Identify rows with missing values
missing_values = weather_df[weather_df.isnull().any(axis=1)]

# Display the rows with missing values
print(missing_values)

# Check the total number of missing rows
print(f"Total rows with missing values: {missing_values.shape[0]}")
45

           Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4    2013-12-25         58        50        41           44.0          40.0   
6    2013-12-27         60        53        45           41.0          39.0   
7    2013-12-28         62        51        40           43.0          39.0   
42   2014-02-01         76        66        55           62.0          59.0   
51   2014-02-10         60        48        35           49.0          36.0   
...         ...        ...       ...       ...            ...           ...   
1269 2017-06-12         94        85        75           72.0          68.0   
1288 2017-07-01         98        89        79           76.0          72.0   
1304 2017-07-17         98        88        77           75.0          71.0   
1308 2017-07-21        104        91        77           74.0          68.0   
1313 2017-07-26        103        91        78           75.0          70.0   

      DewPointLowF  HumidityHighPercent  HumidityAv

45

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [31]:
# Your code here
# Count the total number of rows in the original dataset
total_rows = weather_df.shape[0]
print(f"Total number of rows in the dataset: {total_rows}")

# Count the number of rows with missing values
missing_rows = missing_values.shape[0]
print(f"Number of rows with missing values: {missing_rows}")


Total number of rows in the dataset: 1319
Number of rows with missing values: 136


#### Calculate the ratio of missing rows to total rows

In [32]:
# Your code here
# Calculate the ratio of missing rows to total rows
missing_ratio = missing_rows / total_rows
print(f"Ratio of missing rows to total rows: {missing_ratio:.2%}")

Ratio of missing rows to total rows: 10.31%


As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [33]:
# Your code here
# Count missing values per column
missing_per_column = weather_df.isna().sum()

# Display the columns with missing values
print(missing_per_column[missing_per_column > 0])

DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
dtype: int64


#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [34]:
# Your code here
# Calculate the number of missing values in 'PrecipitationSumInches'
missing_precipitation = weather_df['PrecipitationSumInches'].isna().sum()

# Calculate the total number of rows
total_rows = len(weather_df)

# Compute the ratio of missing values
missing_ratio = missing_precipitation / total_rows

# Display the result as a percentage
print(f"Missing values in 'PrecipitationSumInches': {missing_precipitation}")
print(f"Total rows: {total_rows}")
print(f"Missing data ratio: {missing_ratio:.2%}")


Missing values in 'PrecipitationSumInches': 124
Total rows: 1319
Missing data ratio: 9.40%


Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [35]:
# Your code here 

# Drop the column with missing values
weather_df.drop(columns=['PrecipitationSumInches'], inplace=True)

# Print `austin` to confirm the column is indeed removed
print(weather_df.head())


        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0 

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [37]:
# Your code here
# Interpolating missing values and assigning to a new variable
austin_fixed = weather_df.interpolate(inplace=False)

# Display the first few rows to compare
print(austin_fixed.head())


        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0 

C:\Users\navin\AppData\Local\Temp\ipykernel_5944\3295544458.py:3: FutureWarning: DataFrame.interpolate with object dtype is deprecated and will raise in a future version. Call obj.infer_objects(copy=False) before interpolating instead.
  austin_fixed = weather_df.interpolate(inplace=False)


#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [38]:
# Your code here
# Check for missing data in the interpolated dataframe
print("Missing values in austin_fixed:")
print(austin_fixed.isnull().sum().sum())  # Should be 0 if interpolation worked correctly

# Check for missing data in the original dataframe
print("Missing values in weather_df:")
print(weather_df.isnull().sum().sum())  # Should show the original count of missing values

Missing values in austin_fixed:
0
Missing values in weather_df:
80


# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [39]:
# Your code here:
# Check the unique values and their counts in the 'Events' column
print(weather_df['Events'].value_counts())

Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64


Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [40]:
# Your answer:
# Check the unique values and their counts in the 'Events' column
print(weather_df['Events'].value_counts())

Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64


#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [41]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

# Your code here
# List of event types
event_types = ['Rain', 'Snow', 'Fog', 'Thunderstorm']

# Create a dummy column for each event with 0 values
for event in event_types:
    austin_fixed[event] = 0

# Print your new dataframe to check whether new columns have been created:

print(austin_fixed.head())

        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  W

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [42]:
# Your code here
# Populate the dummy columns based on the Events column
for event in event_types:
    austin_fixed[event] = austin_fixed['Events'].str.contains(event, na=False).astype(int)

# Display the first few rows to verify the changes
print(austin_fixed[['Events'] + event_types].head())


                Events  Rain  Snow  Fog  Thunderstorm
0  Rain , Thunderstorm     1     0    0             1
1                          0     0    0             0
2                          0     0    0             0
3                          0     0    0             0
4                          0     0    0             0


#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [43]:
# Your code here# Print out the first few rows of austin_fixed to check if the event columns are populated
print(austin_fixed.head())



        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  W

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [45]:
# Your code here
austin_fixed.drop(columns=['Events'], inplace=True)

# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [46]:
# Your code here
# Print the dataframe to confirm the column has been removed
print(austin_fixed.head())

        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLow

#### Print `austin_fixed` to check your `Date` column.

In [47]:
print(austin_fixed.head(5))

        Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  VisibilityHighMiles  VisibilityAvgMiles  VisibilityLow

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [48]:
# Your code here:
# Define predictor variables (X)
X = austin_fixed[['TempHighF',
                  'TempLowF',
                  'DewPointHighF',
                  'DewPointAvgF',
                  'DewPointLowF',
                  'HumidityHighPercent',
                  'HumidityAvgPercent',
                  'HumidityLowPercent',
                  'SeaLevelPressureHighInches',
                  'SeaLevelPressureAvgInches',
                  'SeaLevelPressureLowInches',
                  'VisibilityHighMiles',
                  'VisibilityAvgMiles',
                  'VisibilityLowMiles',
                  'WindHighMPH',
                  'WindAvgMPH',
                  'WindGustMPH',
                  'Snow',
                  'Fog',
                  'Rain',
                  'Thunderstorm']]

# Define response variable (y)
y = austin_fixed[['TempAvgF']]

# Display first few rows to verify
print(X.head())
print(y.head())

   TempHighF  TempLowF  DewPointHighF  DewPointAvgF  DewPointLowF  \
0         74        45           67.0          49.0          43.0   
1         56        39           43.0          36.0          28.0   
2         58        32           31.0          27.0          23.0   
3         61        31           36.0          28.0          21.0   
4         58        41           44.0          40.0          36.0   

   HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0                 93.0                75.0                57.0   
1                 93.0                68.0                43.0   
2                 76.0                52.0                27.0   
3                 89.0                56.0                22.0   
4                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  ...  \
0                       29.86                      29.68  ...   
1                       30.41                      30.13  

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [49]:
#Your code here:
from sklearn.model_selection import train_test_split

Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [50]:
#Your code here:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [51]:
# Your code here:
import math

ts_rows = math.ceil(len(austin_fixed) * 0.8)
print(ts_rows)

1056


Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [52]:
# Your code here:
X_ts_train = X.iloc[:ts_rows]  # First 80% for training
X_ts_test = X.iloc[ts_rows:]   # Remaining 20% for testing

print(X_ts_train.shape, X_ts_test.shape)

(1056, 21) (263, 21)


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [53]:
# Your code here:
y_ts_train = y.iloc[:ts_rows]  # First 80% for training
y_ts_test = y.iloc[ts_rows:]   # Remaining 20% for testing

print(y_ts_train.shape, y_ts_test.shape)

(1056, 1) (263, 1)
